homework2 submission

faridaeid opened the pull request.

ssancho2 commented on HW2Question1.cpp:
```diff
@@ -0,0 +1,77 @@
+
+};
+
+bool printAncestors(treeNode *head, int key, vector<int>& answer)
```
Please add a comment to the function, explaining what it does, what is expected from the tree, how it behaves in special cases.

ssancho2 replied:
Also, it's usually better to use pointers when you modify the parameter, this makes it explicit to the caller, consider:
printAncestors(head, key, answer)
vs
printAncestors(head, key, &answer)

ssancho2 replied:
Also, the name is a bit misleading, it doesn't print any nodes, or anything :)

ssancho2 commented on HW2Question1.cpp:
```diff
@@ -0,0 +1,77 @@
+                cout<<answer[i]<<endl;
+        }
+        else
+            cout<<"This is the root, no Ancestors"<<endl;
```
No big deal, but most style guides require braces even when there is only one command.

ssancho2 commented on HW2Question1.cpp:
```diff
@@ -0,0 +1,77 @@
+    cin>>key;
+    
+    
+    // constructing thee tree in the assignment handout
```
thee -> the

ssancho2 commented on HW2Question1.cpp:
```diff
@@ -0,0 +1,77 @@
+    treeNode *left;
+    treeNode *right;
+    
+    treeNode(int v):value(v), left(NULL), right(NULL){}
```
Try to use nullptr instead of NULL

ssancho2 commented on HW2Question2.cpp:
```diff
@@ -0,0 +1,91 @@
+    return false;
+}
+
+int LCA(treeNode *head, int key1, int key2)
```
Please use a full name.
Also, add a comment to the function.
It's a bit borderline, but the problem statement says 'Avoid storing additional nodes in a data structure.', storing keys to the nodes in a vector is kind of the same. I think that is to avoid this solution :), if you can, please think of another solution.

ssancho2 commented on HW2Question2.cpp:
```diff
@@ -0,0 +1,91 @@
+
+// find path of each node and checks if the node is in the tree or not
+
+bool findPath(treeNode * head, int key, vector<int>& answer)
```
Same as with previous problem, please use a pointer for the parameter you want to modify.

ssancho2 commented on HW2Question2.cpp:
```diff
@@ -0,0 +1,91 @@
+    if(!findPath(head, key1, path1) || !findPath(head, key2, path2))
+        return -1;
+    
+    // one of the keys is the root then no common ancestors
```
Why not? The common ancestor can be one of the nodes.

ssancho2 commented on HW2Question2.cpp:
```diff
@@ -0,0 +1,91 @@
+        if(path1[i] != path2[i])
+            break;
+    
+    if(path1[i-1] == key1 || path1[i-1] == key2)
```
Again, no need to go one up, the common ancestor can be one of the nodes.

faridaeid commented:
Hi ! I updated question 1 and added the suggestions in the comments. I also added a new file with a new solution for question 2 without using extra data storage

ssancho2 commented on HW2Question2Updatded.cpp:
```diff
@@ -0,0 +1,90 @@
+{
+    if(root == nullptr) return false;
+    if(root->value == key) return true;
+    else return (checkNodeInTree(root->left, key) || checkNodeInTree(root->right, key));
```
Not sure 'else' here adds clarity.

ssancho2 commented on HW2Question2Updatded.cpp:
```diff
@@ -0,0 +1,90 @@
+    else return (checkNodeInTree(root->left, key) || checkNodeInTree(root->right, key));
+}
+
+/// get lowest common ancestor
```
This comment doesn't add much to what the function name already says :)

ssancho2 commented on HW2Question2Updatded.cpp:
```diff
@@ -0,0 +1,90 @@
+
+// Checks if node it is the tree or not
+
+bool checkNodeInTree(treeNode* root, int key)
```
Having this check separated from the lowestCommonAncestor is kind of error prone. Ideally, there is one lowestCommonAncestor called by users, that calls checkNodeInTree and then calls an internal version of lowestCommonAncestor that is not public and can assume nodes are there.

ssancho2 commented on HW2Question2Updatded.cpp:
```diff
@@ -0,0 +1,90 @@
+
+/// get lowest common ancestor
+
+treeNode* lowestCommonAncestor(treeNode *head, int key1, int key2)
```
Check that key1 != key2

ssancho2 commented on HW2Question2Updatded.cpp:
```diff
@@ -0,0 +1,90 @@
+    {
+        cout<<"One or both of the keys are not in the tree"<<endl;
+    }
+
```
remember to delete the nodes.

faridaeid commented:
Hi. I changed the code according to the comments. Waiting for your feedback :) Thanks

ssancho2 commented on HW2Question2Updatded.cpp:
```diff
@@ -54,6 +55,36 @@ treeNode* lowestCommonAncestor(treeNode *head, int key1, int key2)
+    
+    else if(checkNodeInTree(head, key1) && checkNodeInTree(head, key1))
+    {
+        return lowestCommonAncestorPrivate(head, key1, key2);
```
Minor comment about performance: You are potentially exploring the tree three times, which is two times too many :)
Then again performance is not a problem until it is :)

ssancho2 commented on HW2Question2Updatded.cpp:
```diff
@@ -54,6 +55,36 @@ treeNode* lowestCommonAncestor(treeNode *head, int key1, int key2)
+{
+    if(head == nullptr) return;
+    
+    if(head->left != nullptr) deleteNodes(head->left);
```
No need the check, there is already the check on line 80

ssancho2 commented on HW2Question2Updatded.cpp:
```diff
@@ -77,14 +108,11 @@ int main()
+    
+    treeNode* answer = lowestCommonAncestor(head, key1, key2);
+    if(answer != nullptr)
         cout<<"LCA = "<<answer->value<<endl;
```
Maybe add a comment that on case of error the lowestCommonAncestor function already reported the problem.
